### PR TITLE
interfaces,interfaces/types,deamon: mass internal rename to interfaces

### PR DESCRIPTION
**Pull request by zyga**
This branch changes all of the "skill" terminology to the corresponding "interface" terminology. The only exception is things that affect the wire format (those will be changed by a subsequent patch).

Please note that I chose not to rename ``interfaces/types`` directory in this go as a proper directory name was not agreed upon yet. This can be corrected with a later patch as the impact of this change is minimal.

### CODE REVIEW

**Comment by chipaca** [comment]
:+1:

File: skills/repo.go [hunk 1]
```diff
@@ -1,569 +0,0 @@
...
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU General Public License for more details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this program.  If not, see <http://www.gnu.org/licenses/>.
- *
- */
-
-package skills
```
Comment (niemeyer) [inline skills/repo.go]: Please don't move files in the same PR doing changes in those same files, as it gets really hard to review.

File: interfaces/core.go [hunk 2]
```diff
@@ -25,56 +25,56 @@ import (
...
+	// given security system to allow a snap to offer a plug of this interface.
 	//
-	// An empty snippet is returned when the skill doesn't require anything
+	// An empty snippet is returned when the plug doesn't require anything
 	// from the security system to work, in addition to the default
-	// configuration.  ErrUnknownSecurity is returned when the skill cannot
+	// configuration.  ErrUnknownSecurity is returned when the plug cannot
 	// deal with the requested security system.
-	SkillSecuritySnippet(skill *Skill, securitySystem SecuritySystem) ([]byte, error)
+	PlugSecuritySnippet(plug *Plug, securitySystem SecuritySystem) ([]byte, error)
```
Comment (niemeyer) [inline interfaces/core.go]: Unrelated to the rename, but this should also have the slot in. It's symmetrical to the respective slot method.

File: interfaces/core.go [hunk 3]
```diff
@@ -25,56 +25,56 @@ import (
...
 }
 
-// Type describes a group of interchangeable capabilities with common features.
-// Types are managed centrally and act as a contract between system builders,
+// Interface describes a group of interchangeable capabilities with common features.
+// Interfaces are managed centrally and act as a contract between system builders,
 // application developers and end users.
-type Type interface {
-	// Unique and public name of this type.
+type Interface interface {
```
Comment (niemeyer) [inline interfaces/core.go]: The Interface interface.. ugh.. not much to do.

File: interfaces/core.go [hunk 4]
```diff
@@ -25,56 +25,56 @@ import (
...
+	Interface string
+	Attrs     map[string]interface{}
+	Apps      []string
+	Label     string
 }
 
-// Type describes a group of interchangeable capabilities with common features.
-// Types are managed centrally and act as a contract between system builders,
+// Interface describes a group of interchangeable capabilities with common features.
+// Interfaces are managed centrally and act as a contract between system builders,
```
Comment (niemeyer) [inline interfaces/core.go]: s/are managed centrally and//

That became invalid once we expanded the scope moving out of "interface type". People will manage interfaces on their own snappy systems.

File: interfaces/repo.go [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+		return fmt.Errorf("cannot remove plug %q from snap %q, it is still connected", plugName, snapName)
+	}
+	delete(r.plugs[snapName], plugName)
+	if len(r.plugs[snapName]) == 0 {
+		delete(r.plugs, snapName)
+	}
+	return nil
+}
+
+// AllSlots returns all plug slots of the given interface.
```
Comment (niemeyer) [inline interfaces/repo.go]: plug slots?  There are several of those.

Can you please do a pass of self-review to make sure the automated renames you've done ended up well?

File: interfaces/repo.go [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+	}
+	// Reject plug with invalid names
+	if err := ValidateName(plug.Name); err != nil {
+		return err
+	}
+	i := r.ifaces[plug.Interface]
+	if i == nil {
+		return fmt.Errorf("cannot add plug, interface %q is not known", plug.Interface)
+	}
+	// Reject plug that don'i pass interface-specific sanitization
```
Comment (niemeyer) [inline interfaces/repo.go]: s/'i/'t

File: interfaces/repo.go [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+		return fmt.Errorf("cannot disconnect plug from slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	for plug := range r.slotPlugs[slot] {
+		r.disconnect(plug, slot)
+	}
+	return nil
+}
+
+// disconnectPlugFromPlugSlot finds a specific plug slot and plug and disconnects it.
+func (r *Repository) disconnectPlugFromPlugSlot(plugSnapName, plugName, slotSnapName, slotName string) error {
```
Comment (niemeyer) [inline interfaces/repo.go]: PlugSlot

File: interfaces/repo.go [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+	for _, slot := range r.slots[slotSnapName] {
+		for plug := range r.slotPlugs[slot] {
+			r.disconnect(plug, slot)
+		}
+	}
+	return nil
+}
+
+// disconnectEverythingFromPlugSlot finds a specific plug slot and disconnects all the plugs connected there.
+func (r *Repository) disconnectEverythingFromPlugSlot(slotSnapName, slotName string) error {
```
Comment (niemeyer) [inline interfaces/repo.go]: PlugSlot

File: interfaces/repo.go [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+	// Ensure that slot and plug are connected
+	if !r.slotPlugs[slot][plug] {
+		return fmt.Errorf("cannot disconnect plug %q from snap %q from slot %q from snap %q, it is not connected",
+			plugName, plugSnapName, slotName, slotSnapName)
+	}
+	r.disconnect(plug, slot)
+	return nil
+}
+
+// disconnect disconnects a specific plug from a specific plug slot.
```
Comment (niemeyer) [inline interfaces/repo.go]: `"specific plug from a specific plug slot"` => `"plug from a slot"`

File: interfaces/repo.go [hunk 10]
```diff
@@ -0,0 +1,569 @@
...
+		for plug := range r.slotPlugs[slot] {
+			result[slot] = append(result[slot], plug)
+		}
+		sort.Sort(byPlugSnapAndName(result[slot]))
+	}
+	return result
+}
+
+// ConnectedBy returns all of the plugs connected by a given snap.
+func (r *Repository) ConnectedBy(snapName string) map[*Plug][]*Slot {
```
Comment (niemeyer) [inline interfaces/repo.go]: ConnectedBy doesn't sound right anymore.

Let's rename the pair to:

ConnectedTo => ConnectedSlots
ConnectedBy => ConnectedPlugs

**Comment by niemeyer** [comment]
Some comments, and it felt hard not to zone out after a some time. I expect we'll need to do further terminology fixes soon, but it's probably best to get this in sooner and have smaller branches addressing other issues.

Again, please don't move files with changes in the future. It turns an already large review into something much larger and harder to review.

LGTM once you're happy about these points.

File: interfaces/repo.go (hunk 5 above)
Comment (zyga) [inline interfaces/repo.go]: Yes, sorry. I did try to fix those but I apparently miss a few.

File: interfaces/core.go (hunk 2 above)
Comment (zyga) [inline interfaces/core.go]: Noted, added to the TODO list

**Pull request merged**